(DOCS-3777) Update Pivotal Platform instructions (#14955)
DataDog/documentation: content/en/integrations/guide/pivotal-cloud-foundry-manual-setup.md

---
title: Pivotal Cloud Foundry Manual Setup Guide

> kind: guide

Reviewer: If this is a guide, can you update the [Integrations Guides index page](https://docs.datadoghq.com/integrations/guide/)?

description: "Steps for manually setting up the Pivotal Cloud Foundry Integration"

> aliases:
>   - /integrations/pivotal-platform

Reviewer: The alias for the page already exists in `content/en/integrations/vmware_tanzu_application_service.md`.

```suggestion
```

Author: Thanks for catching this!

further_reading:
- link: "https://www.datadoghq.com/blog/monitor-tanzu-application-service/"
  tag: "Blog"
  text: "Monitor applications running on VMware Tanzu Application Service"
---

## Overview

> Pivotal Cloud Foundry (PCF) deployments can send metrics and events to Datadog. You can track the health and availability of all nodes in a deployment, monitor the jobs they run, collect metrics from the Loggregator Firehose, and more. Use this page to manually set up monitoring for your PCF application.

Reviewer: ```suggestion
Pivotal Cloud Foundry (PCF) deployments can send metrics and events to Datadog. You can track the health and availability of all nodes in a deployment, monitor the jobs they run, collect metrics from the Loggregator Firehose, and more. This page walks you through how to manually set up monitoring for your PCF application.
```

> There are three main components for the PCF integration with Datadog. First, the buildpack is used to collect custom metrics from your applications. Second, the BOSH Release collects metrics from the platform. Third, the Loggregator Firehose Nozzle collects all other metrics from your infrastructure. Read the [Datadog VMware Tanzu Application Service architecture][32] guide for more information.

Reviewer: ```suggestion
There are three main components for the PCF integration with Datadog. First, the buildpack is used to collect custom metrics from your applications. Second, the BOSH Release collects metrics from the platform. Third, the Loggregator Firehose Nozzle collects all other metrics from your infrastructure. For more information, see the [Datadog VMware Tanzu Application Service architecture guide][32].
```

Author: For this, I think it's better to use `read` instead of `see`, as it implies a more focused attention. Would you be open to modifying your suggestion to incorporate `read`?

## Monitor your applications

Use the **Datadog Pivotal Cloud Foundry Buildpack** to monitor your PCF application. This is a [supply buildpack][2] for PCF that installs the Datadog Container Agent, Datadog Trace Agent for APM, and Datadog DogStatsD binary file in the container your app is running on.

### Pivotal Cloud Foundry < 1.12

The Datadog buildpack uses the Pivotal Cloud Foundry [Pushing an App with Multiple Buildpacks][3] feature that was introduced in version `1.12`.

For older versions, Pivotal Cloud Foundry provides a backwards compatible version of this feature in the form of a [multi-buildpack][4]. You must install and configure this version in order to use Datadog's buildpack.

> 1. **Upload the multi-buildpack back-port.** Download the latest multi-buildpack release and upload it to your Pivotal Cloud Foundry environment.

Reviewer: ```suggestion
1. Upload the multi-buildpack back-port. Download the latest multi-buildpack release and upload it to your Pivotal Cloud Foundry environment.
```

    ```shell
    cf create-buildpack multi-buildpack ./multi-buildpack-v-x.y.z.zip 99 --enable
    ```

> 2. **Add a multi-buildpack manifest to your application.** As detailed in the [usage section][5] of the multi-buildpack repository, create a `multi-buildpack.yml` file at the root of your application and configure it for your environment. Add a link to the Datadog Pivotal Cloud Foundry Buildpack and to your regular buildpack:

Reviewer: ```suggestion
2. Add a multi-buildpack manifest to your application. As detailed in the [usage section][5] of the multi-buildpack repository, create a `multi-buildpack.yml` file at the root of your application and configure it for your environment. Add a link to the Datadog Pivotal Cloud Foundry Buildpack and to your regular buildpack:
```

    ```yaml
    buildpacks:
      - "https://cloudfoundry.datadoghq.com/datadog-cloudfoundry-buildpack/datadog-cloudfoundry-buildpack-3.1.0.zip"
      - "https://github.com/cloudfoundry/ruby-buildpack#v1.7.18" # Replace this with your regular buildpack
    ```

      The URLs for the Datadog Buildpack are:

      - `https://cloudfoundry.datadoghq.com/datadog-cloudfoundry-buildpack/datadog-cloudfoundry-buildpack-latest.zip`
      - `https://cloudfoundry.datadoghq.com/datadog-cloudfoundry-buildpack/datadog-cloudfoundry-buildpack-x.y.z.zip`

      Do not use the `latest` version here (replace `x.y.z` with the specific version you want to use).

>       **Important**: Your regular buildpack should be the last in the manifest to act as a final buildpack. To learn more see [Pivotal Cloud Foundry's How Buildpacks Work][6].

Reviewer: ```suggestion
      **Your regular buildpack should be the last in the manifest to act as a final buildpack**. To learn more, see [Pivotal Cloud Foundry's How Buildpacks Work][6].
```

> 3. **Push your application with the multi-buildpack**. Ensure that the `multi-buildpack` is the buildpack selected by Pivotal Cloud Foundry for your application:

Reviewer: ```suggestion
3. Push your application with the multi-buildpack. Ensure that the `multi-buildpack` is the buildpack selected by Pivotal Cloud Foundry for your application:
```

    ```shell
    cf push <YOUR_APP> -b multi-buildpack
    ```

### Pivotal Cloud Foundry Platform >= 1.12

> 1. **Upload the Datadog Pivotal Cloud Foundry Buildpack.** [Download the latest Datadog build pack release][7] and upload it to your Pivotal Cloud Foundry environment.

Reviewer: ```suggestion
1. Upload the Datadog Pivotal Cloud Foundry Buildpack. [Download the latest Datadog build pack release][7] and upload it to your Pivotal Cloud Foundry environment.
```

    ```shell
    cf create-buildpack datadog-cloudfoundry-buildpack ./datadog-cloudfoundry-buildpack-latest.zip
    ```

> 2. **Push your application with the Datadog buildpack and your buildpacks.** The process to push your application with multiple buildpacks is described in [Pushing an App with Multiple Buildpacks][3].

Reviewer: ```suggestion
2. Push your application with the Datadog buildpack and your buildpacks. The process to push your application with multiple buildpacks is described in [Pushing an App with Multiple Buildpacks][3].
```

    ```shell
    cf push <YOUR_APP> --no-start -b binary_buildpack
    cf v3-push <YOUR_APP> -b datadog-cloudfoundry-buildpack -b <YOUR-BUILDPACK-1> -b <YOUR-FINAL-BUILDPACK>
    ```

>       **Important**: If you were using a single buildpack before, it should be the last one loaded so it acts as a final buildpack. To learn more see [Pivotal Cloud Foundry's How Buildpacks Work][6].

Reviewer: ```suggestion
      **If you were using a single buildpack before, it should be the last one loaded so it acts as a final buildpack**. To learn more, see [Pivotal Cloud Foundry's How Buildpacks Work][6].
```

      generate_processes: true    # to enable the process check
```

To see which `datadog-agent` release version was uploaded earlier, run `bosh releases`.

#### Load the runtime.yml

Check if you have a previously configured `runtime-config` by running:

```text
# BOSH CLI v1
`bosh runtime-config`
# BOSH CLI v2
bosh -e <BOSH_ENV> runtime-config
```

In BOSH v2, if the `runtime.yml` file is empty, you should see the response: `No runtime config`.

#### Enable extra Agent checks

For each extra Agent check to enable across your deployment, add its configuration under the `properties.dd.integrations` key, for example:

```yaml
properties:
    dd:
        integrations:
            directory:
                init_config: {}
                instances:
                    directory: '.'
            #process:
            #  init_config: {}
            #...
```

The configuration under each check name uses the same format as when configuring the check in its own file in the Agent's `conf.d` directory.

Everything you configure in `runtime.yml` applies to every node. You cannot configure a check for a subset of nodes in your deployment.

> To customize configuration for the default checks -- system, network, disk, and NTP -- see the [full list of configuration options][19] for the Datadog Agent BOSH release.

Reviewer: ```suggestion
To customize configuration for the default checks (system, network, disk, and NTP), see the [full list of configuration options][19] for the Datadog Agent BOSH release.
```

#### Sync the runtime configuration to the BOSH Director

```text
# BOSH CLI v1
bosh update runtime-config runtime.yml
# BOSH CLI v2
bosh update-runtime-config -e <BOSH_ENV> runtime.yml
```

#### Redeploy your PCF deployment

```text
# BOSH CLI v1
bosh deployment <YOUR_DEPLOYMENT_MANIFEST>.yml
bosh -n deploy --recreate
# BOSH CLI v2
bosh -n -d <YOUR_DEPLOYMENT> -e <BOSH_ENV> deploy --recreate <YOUR_DEPLOYMENT_MANIFEST>.yml
```

Since runtime configuration applies globally, BOSH redeploys every node in your deployment. If you have more than one deployment, redeploy all deployments to install the Datadog Agent everywhere.

#### Verify the Agent is installed everywhere

> To check if the Agent installations were successful, filter by `cloudfoundry` on the [Host map][20] in Datadog. The Datadog Agent BOSH release tags each host with `cloudfoundry`. Optionally, group hosts by any tag, such as `bosh_job`, as in the following screenshot:

Reviewer: ```suggestion
To check if the Agent installations were successful, filter by `cloudfoundry` on the [Host Map][20]. The Datadog Agent BOSH release tags each host with `cloudfoundry`. Optionally, group hosts by any tag, such as `bosh_job`, as in the following screenshot:
```

    ...
```

#### Integration configurations discovery
The Datadog Cluster Agent discovers integrations based on an `AD_DATADOGHQ_COM` environment variable set in your applications.
This environment variable is a JSON object containing the Autodiscovery configuration templates for your application. The Datadog Cluster Agent can discover and render two types of configurations:
  1. Configurations for services bound to your application, whether they be user-provided or from a service broker.
  2. Configurations for services running inside your application, for example, a web-server.

The JSON object should be a dictionary associating a service name to its Autodiscovery template:
```
{
    "<SERVICE_NAME>": {
        "check_names": [<LIST_OF_INTEGRATION_NAMES_TO_CONFIGURE>],
        "init_configs": [<LIST_OF_INIT_CONFIGS>],
        "instances": [<LIST_OF_INSTANCES>],
        "variables": [<LIST_OF_VARIABLES_DEFINITIONS>]
    }
}
```

For services bound to the application, the `<SERVICE_NAME>` should be the name of the service as it appears in the `cf services` command output. For services running inside the application, the `<SERVICE_NAME>` can be anything.  

> The `variables` key is used only for bound services to resolve template variables inside the configuration template, and must contain the JSON path of the desired value for the `VCAP_SERVICES` environment variable. You can inspect this with command `cf env <APPLICATION_NAME>`.

Reviewer: ```suggestion

The `variables` key is used only for bound services to resolve template variables inside the configuration template, and must contain the JSON path of the desired value for the `VCAP_SERVICES` environment variable. You can inspect this with the `cf env <APPLICATION_NAME>` command.
```

                "timeout": 1
            }
        ]
    }
    "postgres-service-name": {
        "check_names": ["postgres"],
        "init_configs": [{}],
        "instances": [
            {
                "host": "%%host%%",
                "port": 5432,
                "username": "%%username%%",
                "dbname": "%%dbname%%",
                "password": "%%password%%"
            }
        ],
        "variables": {
            "host": "$.credentials.host",
            "username": "$.credentials.Username",
            "password": "$.credentials.Password",
            "dbname": "$.credentials.database_name"
        }
    }
}'
```

This example demonstrates the accompanying `VCAP_SERVICES` environment variable:

```
VCAP_SERVICES: '{
    "my-postgres-service": [
        {
            "credentials": {
                Password: "1234",
                Username: "User1234",
                host: "postgres.example.com",
                database_name: "my_db",
            },
            "name": "postgres-service-name",
        }
    ]
}'
```

In the example above, the first item `web_server` is a configuration for a service running inside the application.

> There are no `variables`, and it uses the template variables `%%host%%` and `%%port%%` available through Autodiscovery.

Reviewer: ```suggestion
There are no `variables`, and it uses the `%%host%%` and `%%port%%` template variables available through Autodiscovery.
```

The second item `postgres-service-name` is a configuration for a service bound to the application.
To resolve the template variables, it uses the `variables` dictionary to define the values used in the instance configuration.
This dictionary contains a JSONPath object indicating where to find the variable values for the service `postgres-service-name` defined in the `VCAP_SERVICES` environment variable.

#### Improve CCCache performance on cache miss

For Datadog Agent versions `7.40.1` and later, you can add more flags to increase control over the CCCache behavior and the number of API calls:

- `refresh_on_cache_miss` to control cache miss behavior
- Split `advanced_tags` into `sidecars_tags` and `isolation_segments_tags`

#### Improved tagging for application containers and process discovery

> Once the two releases are linked, the Datadog Cluster Agent automatically provides cluster-level metadata, which the node Agents attach as tags to their corresponding Cloud Foundry application containers.

Reviewer: ```suggestion
Once the two releases are linked, the Datadog Cluster Agent automatically provides cluster-level metadata, which the Node Agents attach as tags to their corresponding Cloud Foundry application containers.
```

      subscription_id: datadog-nozzle
      # for development only
      # disable_access_control: true
      # for development only; enable if your UAA does not use a verifiable cert
      # insecure_ssl_skip_verify: true
    uaa:
      client: datadog-firehose-nozzle # client name you just configured
      client_secret: "<SECRET_YOU_JUST_CONFIGURED>"
      url: <UAA_URL> # for example, https://uaa.your-cf-domain.com:8443
```

To see all available configuration options, check the [Datadog Firehose Nozzle repository][22].

In the same manifest, add the Datadog Nozzle release name and version:

```yaml
releases:
    # - name: "<SOME_OTHER_RELEASE>"
    #   version: <x.y.z>
    # ...
    - name: datadog-firehose-nozzle
      version: '<VERSION_YOU_UPLOADED>' # specify the real version (x.y.z not 'latest')
```

To see which `datadog-firehose-nozzle` release version was uploaded earlier, run `bosh releases`.

#### Redeploy the deployment

```text
# BOSH CLI v1
bosh deployment cf-manifest.yml
bosh -n deploy --recreate
# BOSH CLI v2
bosh -n -d cf-manifest -e <BOSH_ENV> deploy --recreate cf-manifest.yml
```

#### Verify the Firehose Nozzle is collecting data

> On the [Metrics explorer][23] page in Datadog, search for metrics beginning with `cloudfoundry.nozzle`:

Reviewer: ```suggestion
In the [Metrics Explorer][23], search for metrics beginning with `cloudfoundry.nozzle`.
```
